src/algorithms/reco/LGADHitClustering.cc

```diff
@@ -0,0 +1,141 @@
+
+  double g[3], l[3];
+  l[0] = locPos.a * dd4hep::mm;
+  l[1] = locPos.b * dd4hep::mm;
+  l[2] = 0;
+  nodeMatrix.LocalToMaster(l, g);
+  dd4hep::Position position;
+  position.SetCoordinates(g);
+  return position;
+}
+
+void LGADHitClustering::process(const LGADHitClustering::Input& input,
+                                const LGADHitClustering::Output& output) const {
+  using dd4hep::mm;
+
+  const auto [calibrated_hits] = input;
+  auto [clusters]              = output;
+
+  // collection of ADC values from all sensors and group them by sensor
+  std::unordered_map<dd4hep::rec::CellID, std::vector<edm4eic::TrackerHit>> hitsBySensors;
+
+  for (const auto& calibrated_hit : *calibrated_hits) {
+
+    auto id = calibrated_hit.getCellID();
+    // Get sensors info
+    auto sensorID = this->getSensorInfos(id);
+    hitsBySensors[sensorID].emplace_back(calibrated_hit);
+  }
+
+  for (const auto& sensor : hitsBySensors) {
+    auto cluster = clusters->create();
+    // Right now the clustering algorithm a simple average over all hits in a sensors
+    // Will be problematic near the edges, but it's just an illustration
+    float ave_x = 0, ave_y = 0;
+    double tot_charge = 0;
+    const auto& hits  = sensor.second;
+    if (hits.size() == 0)
+      continue;
+    // find cellID for the cell with maximum ADC value within a sensor
+    auto id            = hits[0].getCellID();
+    auto max_charge    = hits[0].getEdep();
+    auto earliest_time = hits[0].getTime();
+
+    ROOT::VecOps::RVec<double> weights;
+    for (const auto& hit : hits) {
+      // weigh all hits by ADC value
+      auto pos = m_seg->position(hit.getCellID());
+      if (hit.getEdep() < 0)
+        error("Edep for hit at cellID{} is negative. Abort!.", hit.getCellID());
+      ave_x += hit.getEdep() * pos.x();
+      ave_y += hit.getEdep() * pos.y();
+
+      tot_charge += hit.getEdep();
+      if (hit.getEdep() > max_charge) {
+        max_charge = hit.getEdep();
+        id         = hit.getCellID();
+      }
+      earliest_time = std::min(earliest_time, hit.getTime());
+      cluster.addToHits(hit);
+      weights.push_back(hit.getEdep());
+    }
+
+    weights /= tot_charge;
+    ave_x /= tot_charge;
+    ave_y /= tot_charge;
+
+    for (const auto& w : weights)
+      cluster.addToWeights(w);
+
+    edm4eic::Cov3f covariance;
+    edm4hep::Vector2f locPos{static_cast<float>(ave_x / mm), static_cast<float>(ave_y / mm)};
+
+    // CAUTION: surface has to be the cell where the cluster center belongs to
+    // NOT the cell with MAX ADC value
+    const auto* context = m_converter->findContext(id);
+    auto gPos           = this->_local2Global(context, locPos);
+    id                  = m_converter->cellID(gPos);
+
+    if (!id)
+      error("Failed to find cellID for cluster center! Does your detector have hole in the middle "
+            "or somehow concave in shape? This class doesn't work with concave detector.");
+
+    cluster.setSurface(id);
```

Comment:
From my understanding this should be the acts surface id rather than the dd4hep cell id (They might often but not always be the same thing). e.g.
https://github.com/eic/EICrecon/blob/main/src/algorithms/tracking/TrackerMeasurementFromHits.cc#L134